main.py

```diff
@@ -144,62 +145,145 @@ async def nini(ctx):
+
+    for new_rank, name, data in enumerate(all_losers_sorted):
+        if name not in old_ranking:
+            old_rank = len(old_ranking)
+        else:
+            old_rank = old_ranking[name]
+        update = old_rank-new_rank
+        if update > 0:
+            update = "🔼"+str(update)
+        elif update < 0:
+            update = "🔽"+str(-update)
+        else:
+            update = "🟦0"
+        data["rank_update"] = update
+
+    # RANKING MESSAGE
     res = ["Classement du <:nini:696420822855843910>"]
-    res.append("`**{}** | **{}** | **{}** | **{}** | **{}** | **{}**`".format("Pseudo".center(20), "# messages".center(12), "# défaites".center(12),
-                                                      "Ratio".center(7), "Streak".center(12), "Streak max".center(12)))
-    for username, data in all_losers:
-        res.append("`{} | {} | {} | {} | {} | {}`".format(username.center(20), str(data["messages"]).center(12), str(data["errors"]).center(12),
-                                                          str(int(data["messages"]/data["errors"])).center(7), str(data["streak"]).center(12), str(data["streak_max"]).center(12)))
+    res.append("`**{}** | **{}** | **{}** | **{}** | **{}** | **{}** | **{}**`".format("Pseudo".center(20), "# messages".center(12), "# défaites".center(12),
+                                                      "Ratio".center(7), "Streak".center(12), "Streak max".center(12), "Évolution".center(12)))
+
+    for username, data in all_losers_sorted:
+        res.append("`{} | {} | {} | {} | {} | {} | {}`".format(username.center(20), str(data["messages"]).center(12), str(data["errors"]).center(12),
+                                                          str(int(data["messages"]/data["errors"])).center(7), str(data["streak"]).center(12), str(data["streak_max"]).center(12), data["rank_update"].center(12)))
 
     log_info('\n'.join(res))
-    res.append("J'ai lu {} messages !".format(c))
+
+
+    # LOSER MESSAGE AND ROLES
+
+    nLoser = max(NUMBER_OF_LOSER, len(loss_ranking))
+    loss_ranking = loss_ranking[:nLoser]
+
+    if nLoser == 0:
+        loseMsg = "Il n'y à pas de grand loooooser.euse cette semaine."
+    elif nLoser == 1:
+        loseMsg = "Le grand loooooser.euse de cette semaine est "
+    else:
+        loseMsg = "Les {} grands loooooser.euses de cette semaine sont ".format(nLoser)
+
+    for n, l in enumerate(loss_ranking):
+        loseMsg += l[0] + " ({})".format(l[1])
+        if n < nLoser-2:
+            loseMsg += ", "
+        elif n < nLoser-1:
+            loseMsg += " et "
+        else:
+            loseMsg += "."
+
+    res.append(loseMsg)
+
+    # REMOVE ALL CURRENT LOSERS
+    role = get(ctx.guild.roles, name=LOSER_ROLE_NAME)
+
+    for member in ctx.guild.members:
+        if role in member.roles:
+            await member.remove_roles(role)
+
+    for name, r in loss_ranking:
+        if name in all_losers and "id" in all_losers[name]:
```

Comment:
If someone is in `loss_ranking`, they have to already be in all_losers. Is this line useful? Is it some kind of safeguard?

Reply:
safeguard, i know try/catch is more efficient than that defensive programming but force of habit

Reply:
than defensive*